crates/core/src/metadata_store/providers/objstore/object_store_version_repository.rs

```diff
@@ -0,0 +1,442 @@
+                }
+                assert_eq!(bytes, DELETED_HEADER);
+                self.put_if_tag_matches(key, tag, content).await
+            }
+            Err(e) => Err(VersionRepositoryError::Network(e.into())),
+        }
+    }
+
+    async fn get(&self, key: ByteString) -> Result<TaggedValue, VersionRepositoryError> {
+        let path = Path::from(key.to_string());
+        match self.object_store.get(&path).await {
+            Ok(res) => {
+                let etag = res.meta.e_tag.as_ref().ok_or_else(|| {
+                    VersionRepositoryError::Network("expecting an ETag to be present".into())
+                })?;
+                let tag = etag.to_owned().into();
+                let mut buf = res
+                    .bytes()
+                    .await
+                    .map_err(|e| VersionRepositoryError::Network(e.into()))?;
+                if buf.starts_with(&DELETED_HEADER) {
+                    Err(VersionRepositoryError::NotFound)
+                } else {
+                    let bytes = buf.split_off(EXISTS_HEADER.len());
+                    Ok(TaggedValue { bytes, tag })
+                }
+            }
+            Err(Error::NotFound { .. }) => Err(VersionRepositoryError::NotFound),
+            Err(e) => Err(VersionRepositoryError::Network(e.into())),
+        }
+    }
+
+    async fn put_if_tag_matches(
+        &self,
+        key: ByteString,
+        expected: Tag,
+        new_content: Bytes,
+    ) -> Result<Tag, VersionRepositoryError> {
+        let etag = expected.as_string();
+        let update_version = UpdateVersion {
+            e_tag: Some(etag),
+            version: None,
+        };
+
+        let path = Path::from(key.to_string());
+        match self
+            .object_store
+            .put_opts(
+                &path,
+                PutPayload::from_iter([EXISTS_HEADER, new_content]),
+                PutOptions::from(PutMode::Update(update_version)),
+            )
+            .await
+        {
+            Ok(res) => {
+                let etag = res
+                    .e_tag
+                    .ok_or_else(|| VersionRepositoryError::Network("expecting an etag".into()))?;
+                let tag: Tag = etag.into();
+                Ok(tag)
+            }
+            Err(Error::Precondition { .. }) => Err(VersionRepositoryError::PreconditionFailed),
+            Err(e) => Err(VersionRepositoryError::Network(e.into())),
+        }
+    }
+
+    async fn put(
+        &self,
+        key: ByteString,
+        new_content: Bytes,
+    ) -> Result<Tag, VersionRepositoryError> {
+        let path = Path::from(key.to_string());
+        match self
+            .object_store
+            .put(&path, PutPayload::from_iter([EXISTS_HEADER, new_content]))
+            .await
+        {
+            Ok(res) => {
+                let etag = res
+                    .e_tag
+                    .ok_or_else(|| VersionRepositoryError::Network("expecting an etag".into()))?;
```

Comment:
Is a missing `e_tag` something we can recover from? The error variant `Network` sounds as if it should be retryable.